state-chain/pallets/cf-lending-pools/src/chp_lending.rs

```diff
@@ -0,0 +1,459 @@
+}
+
+fn usdc_collateral_required<T: Config>(asset: Asset, loan_principal: AssetAmount) -> AssetAmount {
+	let config = ChpConfig::<T>::get();
+
+	let usdc_loan_value = usdc_equivalent_amount::<T>(asset, loan_principal);
+
+	usdc_loan_value + config.overcollateralisation_target * usdc_loan_value
+}
+
+pub fn swap_for_chp<T: Config>(
+	from_asset: Asset,
+	amount: AssetAmount,
+	to_asset: Asset,
+	loan_id: ChpLoanId,
+) {
+	T::SwapRequestHandler::init_swap_request(
+		from_asset,
+		amount,
+		to_asset,
+		SwapRequestType::Regular { output_action: SwapOutputAction::CreditLendingPool { loan_id } },
+		Default::default(), // broker fees
+		None,               // no refund
+		None,               // no dca parameters
+		SwapOrigin::Internal,
+	);
+}
+
+fn initiate_loan_fees_swap<T: Config>(loan: &mut ChpLoan<T>) {
+	if loan.status != LoanStatus::Finalising {
+		loan.status = LoanStatus::Finalising;
+		swap_for_chp::<T>(COLLATERAL_ASSET, loan.fees_collected_usdc, loan.asset, loan.loan_id);
+	} else {
+		unreachable!();
+	}
+}
+
+fn initiate_soft_liquidation<T: Config>(loan: &mut ChpLoan<T>) {
+	if loan.status == LoanStatus::Active {
+		loan.status = LoanStatus::SoftLiquidation { usdc_collateral: loan.usdc_collateral };
+
+		swap_for_chp::<T>(COLLATERAL_ASSET, loan.usdc_collateral, loan.asset, loan.loan_id);
+
+		loan.usdc_collateral = 0;
+	} else {
+		unreachable!()
+	}
+}
+
+pub fn process_interest_for_loan<T: Config>(
+	current_block: BlockNumberFor<T>,
+	loan: &mut ChpLoan<T>,
+) -> Weight {
+	if current_block.saturating_sub(loan.created_at_block) % INTEREST_PAYMENT_INTERVAL.into() ==
+		0u32.into()
+	{
+		for ChpPoolContribution { principal, .. } in &loan.pool_contributions {
+			let interest_amount_in_loan_asset =
+				(loan.interest_rate.int_mul(INTEREST_PAYMENT_INTERVAL)) * *principal;
+
+			// NOTE: we use asset price at the time the loan was created to make sure
+			// that interest payments are consistent/predictable.
+			let usdc_interest_amount = cf_amm_math::output_amount_ceil(
+				interest_amount_in_loan_asset.into(),
+				loan.asset_price_at_creation,
+			)
+			.unique_saturated_into();
+
+			loan.usdc_collateral.saturating_reduce(usdc_interest_amount);
+			loan.fees_collected_usdc.saturating_accrue(usdc_interest_amount);
+		}
+	}
+
+	Weight::zero()
+}
+
+// Sweeping but it is a no-op if it fails for whatever reason
+fn try_sweep<T: Config>(account_id: &T::AccountId) {
+	use frame_support::sp_runtime::TransactionOutcome;
+
+	storage::with_transaction_unchecked(|| {
+		if T::PoolApi::sweep(account_id).is_ok() {
+			TransactionOutcome::Commit(())
+		} else {
+			TransactionOutcome::Rollback(())
+		}
+	})
+}
+
+pub fn process_collateral_topup<T: Config>(
+	loan: &mut ChpLoan<T>,
+	config: &ChpConfiguration,
+) -> Weight {
+	if loan.overcollateralisation_ratio() < config.overcollateralisation_topup_threshold {
+		let collateral_required =
+			usdc_collateral_required::<T>(loan.asset, loan.total_principal_amount());
```

Comment:
We should be consistent here regarding what config we use. 
I think we should be passing the config through here